### PR TITLE
CPP: Improve TaintedAllocationSize.ql

### PR DESCRIPTION
**Pull request by geoffw0**
This improves the `TaintedAllocationSize.ql` query such that it will report more results on LGTM (about 20 over 38 projects I tested) and catch https://github.com/gsingh93/lgtm-test/blob/master/test.cpp line 47.  

This is another query that should be upgraded to the new taint library, when we've got that to a genuinely usable state (I have some ideas about that but should probably work on the buffer overflow stuff first).

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql [hunk 1]
```diff
@@ -14,15 +14,22 @@
...
+predicate taintedAllocSize(Expr e, Expr source, string taintCause) {
+  (
+    isAllocationExpr(e) or
+    any(MulExpr me | me.getAChild() instanceof SizeofOperator) = e
+  ) and exists(Expr tainted |
+    tainted = e.getAChild() and
+    tainted.getType().getUnspecifiedType() instanceof IntegralType and
+    isUserInput(source, taintCause) and
+    tainted(source, tainted)
+  )
```
Comment (jbj) [inline cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql]: I couldn't parse this predicate body by looking at the left column only. Please autoformat.

File: cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+	MyStruct *arr1 = (MyStruct *)malloc(sizeof(MyStruct)); // GOOD
+	MyStruct *arr2 = (MyStruct *)malloc(tainted); // BAD
+	MyStruct *arr3 = (MyStruct *)malloc(tainted * sizeof(MyStruct)); // BAD
+	MyStruct *arr4 = (MyStruct *)malloc(getTainted() * sizeof(MyStruct)); // BAD [NOT DETECTED]
+	MyStruct *arr5 = (MyStruct *)malloc(sizeof(MyStruct) + tainted); // BAD [NOT DETECTED]
+
+	int size = tainted * 8;
+	char *chars1 = (char *)malloc(size); // BAD
+	char *chars2 = new char[size]; // BA
```
Comment (jbj) [inline cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp]: Missing `D`

Comment (geoffw0) [reply cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/TaintedAllocationSize/test.cpp]: Fixed.

File: cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-190/TaintedAllocationSize.ql]: Done, though it makes little difference to that predicate.

**Comment by geoffw0** [comment]
> I'd like to see the CPP-Differences for this PR.

https://jenkins.internal.semmle.com/job/Query-Changes/job/CPP-Differences/523/ (in progress)

**Comment by geoffw0** [comment]
Diff results are ready (on the link above).

It looks like the `strlen` changes in `TaintTracking.qll` have caused quite a number of results to disappear.  I've only looked at the changes in Boost so far, and they do appear to be what I was aiming for - no longer considering the length of a tainted string to be tainted.  This objective is debatable, but my current feeling is that in many of these situations the length is constrained anyway and so not really tainted, or the use of that length is just to index the string, which is fine.  Or, in the case of `TaintedAllocationSize.ql`, there might be results when a copy of the string is made, which also seems to be safe.  Possibly the results lost for `ArithmeticTainted.ql` are more likely to have been good (e.g. if you double the string length it could perhaps overflow).

**Comment by jbj** [comment]
For `ArithmeticTainted.ql`, even adding 1 to a string length could overflow, especially if it's stored in an `int` variable first. It would be a shame to lose those results, although I'm not sure whether that query is important to us.

Will we lose any results on the Juliet test suite with this PR?

**Comment by geoffw0** [comment]
> Will we lose any results on the Juliet test suite with this PR?

I've started https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp/425/ to confirm.

**Comment by geoffw0** [comment]
Success!  The samate job passed, which means none of the Juliet test suite results have changed.

**Comment by geoffw0** [comment]
One of the query tests was failing due to three false positive results that no longer occur, so I've rebased and added a commit to accept those changes.  I'm not sure what caused them exactly, perhaps a combination of the changes here and some recent changes to the taint library might have fixed these results together?

**Comment by jbj** [comment]
A test is failing.

**Comment by geoffw0** [comment]
Yep, the test failure looks like something we'd expect, except that it's in one of the tests in the other repo.  I've opened https://git.semmle.com/Semmle/code/pull/30782 to fix it.  I've also rebased this PR fixing the trivial conflict in the change notes file.

**Comment by geoffw0** [comment]
Rebased again.

**Comment by jbj** [comment]
The change note needs to be moved to a new file for 1.21 before we can merge this into `master`.

**Comment by geoffw0** [comment]
I've updated the change note.

**Pull request merged**